fvm/Grid/NumericBRadialGridGenerator.cpp

```diff
@@ -0,0 +1,505 @@
+				),
+			Bphi = gsl_spline2d_eval(
+					this->spline_Bphi, this->r[i], 0,
+					this->acc_r, this->acc_theta
+				),
+            psip = gsl_spline_eval_deriv(
+                    this->spline_psi, this->r[i], this->acc_r
+                );
+
+		this->BtorGOverR0[i] = (R/this->Rp) * Bphi;
+        this->psiPrimeRef[i] = psip;
+		//this->psiPrimeRef[i] = (R/this->Rp) * hypot(Br, Bz);
+	}
+	for (len_t i = 0; i < GetNr()+1; i++) {
+		real_t
+			R  = gsl_spline2d_eval(
+					this->spline_R, this->r_f[i], 0,
+					this->acc_r, this->acc_theta
+				),
+			Bphi = gsl_spline2d_eval(
+					this->spline_Bphi, this->r_f[i], 0,
+					this->acc_r, this->acc_theta
+				),
+            psip = gsl_spline_eval_deriv(
+                    this->spline_psi, this->r_f[i], this->acc_r
+                );
+
+		this->BtorGOverR0_f[i] = (R/this->Rp) * Bphi;
+        this->psiPrimeRef_f[i] = psip;
+		//this->psiPrimeRef_f[i] = (R/this->Rp) * hypot(Br, Bz);
+	}
+	
+	rGrid->SetReferenceMagneticFieldData(
+		this->BtorGOverR0, this->BtorGOverR0_f,
+		this->psiPrimeRef, this->psiPrimeRef_f,
+		this->Rp
+	);
+
+    this->isBuilt = true;
+    return true;
+}
+
+/**
+ * Calculate the Jacobian J, normalized to R0, at the specified radius and
+ * poloidal angle.
+ *
+ * r:     Minor radius.
+ * theta: Poloidal angle.
+ *
+ * Optional return parameters:
+ * _R:    Major radius coordinate in the given point.
+ * _dRdt: Poloidal angle derivative of R.
+ * _dZdt: Poloidal angle derivative of vertical coordinate Z.
+ */
+real_t NumericBRadialGridGenerator::JacobianAtTheta(
+    const real_t r, const real_t theta,
+    real_t *_R, real_t *_dRdt, real_t *_dZdt
+) {
+    real_t t = this->_thetaBounded(theta);
+
+    real_t dRdr = gsl_spline2d_eval_deriv_x(this->spline_R, r, t, this->acc_r, this->acc_theta);
+    real_t dRdt = gsl_spline2d_eval_deriv_y(this->spline_R, r, t, this->acc_r, this->acc_theta);
+
+    real_t dZdr = gsl_spline2d_eval_deriv_x(this->spline_Z, r, t, this->acc_r, this->acc_theta);
+    real_t dZdt = gsl_spline2d_eval_deriv_y(this->spline_Z, r, t, this->acc_r, this->acc_theta);
+
+    real_t R    = gsl_spline2d_eval(this->spline_R, r, t, this->acc_r, this->acc_theta);
+
+    if (_R != nullptr) *_R = R;
+    if (_dRdt != nullptr) *_dRdt = dRdt;
+    if (_dZdt != nullptr) *_dZdt = dZdt;
+
+    return R/this->Rp*fabs(dRdr*dZdt - dRdt*dZdr);
+}
+
+/**
+ * Calculate R/R0 at the given poloidal angle.
+ *
+ * r:     Minor radius.
+ * theta: Poloidal angle.
+ */
+real_t NumericBRadialGridGenerator::ROverR0AtTheta(
+    const real_t r, const real_t theta
+) {
+    real_t t = this->_thetaBounded(theta);
+    return gsl_spline2d_eval(this->spline_R, r, t, this->acc_r, this->acc_theta) / this->Rp;
+}
+
+/**
+ * Calculate |\nabla r|^2 at the given poloidal angle.
+ *
+ * r:     Minor radius.
+ * theta: Poloidal angle.
+ */
+real_t NumericBRadialGridGenerator::NablaR2AtTheta(
+    const real_t r, const real_t theta
+) {
+	if (r == 0)
+		return 1.0;
```

Comment:
@Embreus This line only works in the very special case of kappa=1. Any idea about how nabla-r could be evaluated in r=0 in numerical geometry?

Reply:
annoyingly, I don't. However, does anything depend on the value coefficients take in r=0? Since the jacobian is 0 in this point, I imagine that all contributions from this point naturally vanish. Not the most comforting answer... Perhaps extrapolate from two small non-zero radii?

Reply:
Good point! Extrapolation would probably be a bit tricky (not super, but...), and if the value is anyway never going to be used it feels like a waste of time. Should add a comment though...